### PR TITLE
Added eclipse-wtp 3.15.2 to Eclipse (as pseudo Eclipse version 4.13.1).

### LINKED ISSUE #492: XML formatter config is sometimes not applied when running gradle --parallel
At this moment, don't know if this is a gradle or spotless-plugin problem.

Sometimes, when running `gradle --parallel`, the `configFile()` setting is not applied, or is applied too late (after formatting has started in another thread)

Command line used from Jenkins is: `gradle --parallel build check --info`

```
05:26:04 ------------------------------------------------------------
05:26:04 Gradle 5.6.3
05:26:04 ------------------------------------------------------------
05:26:04 
05:26:04 Build time:   2019-10-18 00:28:36 UTC
05:26:04 Revision:     bd168bbf5d152c479186a897f2cea494b7875d13
05:26:04 
05:26:04 Kotlin:       1.3.41
05:26:04 Groovy:       2.5.4
05:26:04 Ant:          Apache Ant(TM) version 1.9.14 compiled on March 12 2019
05:26:04 JVM:          1.8.0_231 (Oracle Corporation 25.231-b11)
05:26:04 OS:           Linux 3.10.0-1062.el7.x86_64 amd64
...
05:26:18 Downloading https://plugins.gradle.org/m2/com/diffplug/spotless/spotless-plugin-gradle/3.26.1/spotless-plugin-gradle-3.26.1.jar to XXXX
05:26:25 Downloading https://plugins.gradle.org/m2/com/diffplug/spotless/spotless-eclipse-base/3.2.1/spotless-eclipse-base-3.2.1.pom to XXXX
05:26:25 Downloading https://plugins.gradle.org/m2/com/diffplug/spotless/spotless-eclipse-wtp/3.15.1/spotless-eclipse-wtp-3.15.1.pom to XXXX
```

This is the spotless version: 
```
classpath group: 'com.diffplug.spotless', name: 'spotless-plugin-gradle', version: '3.26.1'
```

This is the spotless configuration:
```
  spotless {
        java {
            googleJavaFormat('1.7')
            licenseHeaderFile file("$project.rootProject.projectDir/config/license.java")
        }
        format 'gradle', {
            target '**/*.gradle'
            trimTrailingWhitespace()
            indentWithSpaces(2)
            endWithNewline()
        }
        format 'xml', {
            target fileTree('.') {
                include '**/*.xml'
                exclude '**/build/**'
            }
            eclipseWtp('xml').configFile("$project.rootProject.projectDir/config/spotless.xml.prefs")
        }
    }
```

The xml prefs are the standard eclipse prefs:
```
user@host> cat config/spotless.xml.prefs
eclipse.preferences.version=1
indentationChar=space
indentationSize=2
lineWidth=200
user@host>
```

Error message (trimmed):
```
05:26:32 FAILURE: Build failed with an exception.
05:26:32 
05:26:32 * What went wrong:
05:26:32 Execution failed for task ':some-proj:spotlessXml'.
05:26:32 > The following files had format violations:
05:26:32       some-proj/path/some-file.xml
05:26:32           @@ -1,93 +1,117 @@
05:26:32            <?xml·version="1.0"·encoding="UTF-8"·standalone="no"?>
05:26:32           -<tag foo="value that is quite long and is likely to span several lines when formatted">
05:26:32           -··<!--·some comment·-->
05:26:32           -··<child-tag name="some-name"·value="123456789"·/>
05:26:32           +<tag
05:26:32           +\t\tname="value that is quite long and is likely to span several lines when formatted">
05:26:32           +\t\t<!--·some comment·-->
05:26:32           +\t\t<child-tag name="some-name"·value="123456789"·/>
...
```

It should be noted that the file was already formatted (as is seen in the above output, using two spaces) when this error was seen.
Something in either gradle or spotless causes the formatter to start formatting before it has received the correct service config (i.e. the contents of the property file).
I'd be happy to help to figure out what, if you could guide me into where to start looking (e.g. what are the entry points between the spotless plugin/ gradle configuration and the wtp formatter).

### PR DESCRIPTION
**Pull request by fvgh**
Fix #492.
Introduced pseudo Eclipse version 4.13.1 since the new approach does not support multiple WTP configurations (for the same WTP type). So the user still has the possibility to use 4.13.0.

If any user needs multiple WTP configurations, please raise an issue.

### CODE REVIEW

**Comment by fvgh** [comment]
> Hmm, the changelog and documentation for this is a bit tricky. In general, I'm in favor of complete documentation, but in some cases I think it's better to have a descriptive error message, if it makes the documentation shorter and easier to read for the 95% usecase. As Spotless has grown more capable, I think our docs are starting to become a little less beginner-friendly.
> 
Proposals are always welcome.

> The rationale behind the approach we took was that every known user is fine with just one config, so I think it's okay to leave it out of the docs, so long as the changelog and error has the required info to help debug any surprises. Here's another way to do the changelog:
> 
>     * Formerly, the Eclipse-WTP formatter (web tools platform, not java) could encounter errors in parallel multiproject builds.  This is now fixed, but there is a new limitation that WTP must have the same configuration for every subproject it is used in.  For example you cant use `eclipseWtp('xml').configFile('a')` in one project and `eclipseWtp('xml').configFile('b')` in another.

>  You also can't use `eclipseWtp('xml')` in one project and `eclispeWtp('css')` in another.  If this limitation is a problem for your use case, please leave a comment in #538
> 
The second part is incorrect. We already [use different class loader](https://github.com/diffplug/spotless/blob/master/lib-extra/src/main/java/com/diffplug/spotless/extra/EclipseBasedStepBuilder.java#L179) for "different formatter types".
We could use the same option, [as already foreseen](https://github.com/diffplug/spotless/blob/master/lib-extra/src/main/java/com/diffplug/spotless/extra/EclipseBasedStepBuilder.java#L178), also to support multiple configurations.
In this scenario, I would waste some memory to statically cache properties hash sum and select the appropriate `JAR state`. This might of course stock pile old JAR state if the developer changes configuration. So there must be some limit and appropriate error if limit is reached.
So when the documentation get's too messy, maybe we just should adapt the implementation.
> 
> We do generate an error if it is used incorrectly, correct? Ideally we would put a link to #538 into that error, but I think it's also sufficient to just put the stacktrace into the body of #538 so that it is easy to google.

This is an option. This is the [current error](https://github.com/diffplug/spotless/blob/master/_ext/eclipse-wtp/src/main/java/com/diffplug/spotless/extra/eclipse/wtp/sse/PluginPreferences.java#L126). The error was always there for `CSS`, since unlike the other WTP formatters, it does not (or at least I have not found it) provide a way to prevent reloading the configuration from the global WTP plugin preferences.

**Comment by nedtwigg** [comment]
Thanks, I forgot we had built some of that, in that case, maybe this as the changelog entry:

- Formerly, the Eclipse-WTP formatter (web tools platform, not java) could encounter errors in parallel multiproject builds. This is now fixed, but there is a new limitation which probably won't affect any users, but just in case:
  - `eclipseWtp('xml')` in one subproject and `eclipseWtp('css')` in another will work fine, but...
  - `eclipseWtp('xml').configFile('a')` in one subproject and `eclipseWtp('xml').configFile('b')` in another will not, and your build will fail with an exception.  If this limitation is a problem for your use case, please leave a comment in #538.

Except, unless I misunderstand, I actually don't think we have that limitation!  As you pointed out, the `this` in `JarState.getClassLoader(this)` is referring to:

https://github.com/diffplug/spotless/blob/11714b6f9fb0de9b1df75472c8415d2dd100ff5e/lib-extra/src/main/java/com/diffplug/spotless/extra/EclipseBasedStepBuilder.java#L143-L151

so now I think we are already capturing the settings, and there is no limitation.  Am I missing something again?

**Comment by fvgh** [comment]
Mh, and I forgot that we really do a new state per config change. I only remembered the `formatterStepExt` I once introduced....
I don't like it very much, since it adds an unrestricted amount of new objects to a static cache.
Further more it should not be required for most Eclipse formatters.
I put it in as a copy and paste form the [initial JDT formatter ](https://github.com/diffplug/spotless/commit/49790696bbbe6419020a55f3a360fbb04856939d#diff-5a9442ba48f0b2e22bae47802a55acdbL88).
I don't say that it is responsible for problems such as #525, but it definitely seems a waste of memory to me. I have no idea why it was used for the original JDT formatter.

But your are right, we currently have no restrictions on WTP.
I even once [programmed a test]((https://github.com/diffplug/spotless/blob/master/lib-extra/src/test/java/com/diffplug/spotless/extra/wtp/EclipseWtpFormatterStepTest.java#L99-L105)) checking multiple configurations for WTP.

To summarize:

- I will provide a new PR for #492.
- I will open an issue to clarify and cleanup the `EclipseBasedStepBuilder`

**Comment by nedtwigg** [comment]
> I don't like it very much, since it adds an unrestricted amount of new objects to a static cache.
Further more it should not be required for most Eclipse formatters.

For every build tool I'm aware of besides Gradle, the build is a run-once process.  Anything that might be useful is cached, since the whole process is just going to exit soon anyway.

The gradle daemon is super useful, but because it's long-lived there's a chance that the "cache-everything" approach can cause it memory-leak into oblivion.  But I think that's not only okay, but also the right approach for build plugins to take.  The build does not change often, and when it does change restarting the daemon is not a relatively-expensive process anyway.  For the common case where the build does not change then our current approach is the most-efficient, easiest-to-use way to do it.

**Comment by nedtwigg** [comment]
I think this PR is fine as-is, it just doesn't need the doc changes, and can get by with a simpler changelog entry.
